src/Microsoft.ApplicationInsights/Managed/Shared/Extensibility/Implementation/Tracing/HeartbeatProvider.cs

```diff
@@ -0,0 +1,354 @@
+        /// The SDK supplied 'default' payload items to send out with each health heartbeat.
+        /// </summary>
+        private ConcurrentDictionary<string, HeartbeatPropertyPayload> sdkPayloadItems;
+
+        /// <summary>
+        /// The extended payload items to send out with each health heartbeat.
+        /// </summary>
+        private ConcurrentDictionary<string, HeartbeatPropertyPayload> extendPayloadItems;
+
+        private bool disposedValue = false; // To detect redundant calls to dispose
+        private TimeSpan heartbeatInterval; // time between heartbeats emitted specified in milliseconds
+        private TelemetryClient telemetryClient; // client to use in sending our heartbeat
+        private volatile bool isEnabled; // no need for locks or volatile here, we can skip/add a beat if the module is disabled between heartbeats
+
+        public HeartbeatProvider()
+        {
+            this.disabledDefaultFields = null;
+            this.heartbeatInterval = TimeSpan.FromMilliseconds(DefaultHeartbeatIntervalMs);
+            this.extendPayloadItems = new ConcurrentDictionary<string, HeartbeatPropertyPayload>(StringComparer.OrdinalIgnoreCase);
+            this.sdkPayloadItems = new ConcurrentDictionary<string, HeartbeatPropertyPayload>(StringComparer.OrdinalIgnoreCase);
+            this.heartbeatsSent = 0; // count up from construction time
+            this.isEnabled = false; // wait until Initialize is called before this means anything
+        }
+
+        /// <summary>
+        /// Gets the currently defined interval between heartbeats (primarily used in unit testing scenarios)
+        /// </summary>
+        public TimeSpan HeartbeatInterval => this.heartbeatInterval;
+
+        /// <summary>
+        /// Gets or sets the current Instrumentation key being used to send heartbeat data to
+        /// </summary>
+        public string DiagnosticsInstrumentationKey { get; set; }
+
+        /// <summary>
+        /// Gets or sets a value indicating whether or not heartbeats are enabled
+        /// </summary>
+        public bool IsEnabled
+        {
+            get => this.isEnabled;
+            set => this.isEnabled = value;
+        }
+
+        protected Timer HeartbeatTimer { get; set; } // timer that will send each heartbeat in intervals
+
+        public virtual bool Initialize(TelemetryConfiguration configuration, string instrumentationKey = null, TimeSpan? timeBetweenHeartbeats = null, IEnumerable<string> disabledDefaultFields = null, bool isEnabled = true)
+        {
+            if (timeBetweenHeartbeats != null && timeBetweenHeartbeats?.TotalMilliseconds == 0)
+            {
+                return false;
+            }
+
+            if (instrumentationKey != null)
+            {
+                this.DiagnosticsInstrumentationKey = instrumentationKey;
+            }
+
+            if (this.telemetryClient == null)
+            {
+                this.telemetryClient = new TelemetryClient(configuration);
```

Comment:
should you be setting the instrumentation for the telemetryclient to the DiagnosticsInstrumentationKey here? or is that done somewhere else?